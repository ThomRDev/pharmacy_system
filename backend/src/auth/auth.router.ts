import { BaseRouter } from "../shared/router/router"
import { AuthController } from "./controllers/auth.controller"
import { AuthMiddleware } from "./middlewares/auth.middlewares"

export class AuthRouter extends BaseRouter<AuthController,AuthMiddleware>{
  constructor(
    
  ){
    super(AuthController,AuthMiddleware)
  }
  routes(): void {
    this.router.post("/login",
    [ this.middleware.validateRequest.bind(this.middleware) ]
    ,this.controller.login.bind(this.controller))

    this.router.post("/auth/check",[
      this.middleware.validateJWT.bind(this.middleware)
    ],this.controller.checking.bind(this.controller))
  }
}