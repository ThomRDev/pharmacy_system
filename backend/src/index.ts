import { ServerBootstrap } from "./server";

(async ()=>{
  const server = new ServerBootstrap()
  server.config()
  await server.upServer()
})();
