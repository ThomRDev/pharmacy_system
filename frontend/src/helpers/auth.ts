import { pharmacyApi } from "../apis"
import { AxiosError } from "axios"

export const pharmaryLogin = async ({ username,password }:{ username:string,password:string }) => {
  try{
    const { data} = await pharmacyApi.post("/login",{
      username,
      password
    })
    localStorage.setItem("token", data?.accessToken);
    return {
      ...data,
      ...data.employee
    }
  }catch(err){
    localStorage.removeItem("token");
    return (err as AxiosError ).response?.data
  }
}

export const pharmaryCheckAuth = async (token:string) => {
  try{
    const { data} = await pharmacyApi.post("/auth/check",{},{
      headers:{
        "Authorization" : `Bearer ${token}`
      }
    })
    return {
      ...data,
      ...data.data
    }
  }catch(err){
    localStorage.removeItem("token");
    return (err as AxiosError ).response?.data
  }
}