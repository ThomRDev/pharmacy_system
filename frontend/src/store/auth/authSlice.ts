import { createSlice } from '@reduxjs/toolkit';
import type { PayloadAction } from '@reduxjs/toolkit'

export enum AUTH_STATUS {
  CHECKING = 'checking', // revisando
  NOT_AUTHENTICATED = 'not-authenticated', // no esta authenticado
  AUTHENTICATED ='authenticated' // esta authenticado
}

export interface AuthInitialState {
  status       : AUTH_STATUS
  uid          : string | null
  email        : string | null
  username     : string | null
  errorMessage : string | null
  fullname : string | null
  role : string | null
}
const initialState:AuthInitialState = {
  status : AUTH_STATUS.CHECKING,
  uid: null,
  email: null,
  username: null,
  errorMessage: null,
  fullname : null,
  role : null
}
export const authSlice = createSlice({
  name  : 'auth',
  initialState,
  reducers : {
    login : (state:AuthInitialState, action: PayloadAction<AuthInitialState>) => {
      state.status = AUTH_STATUS.AUTHENTICATED
      state.uid = action.payload.uid
      state.email = action.payload.email
      state.username = action.payload.username
      state.errorMessage = null;
      state.fullname = action.payload.fullname
      state.role = action.payload.role
    },
    logout: (state:AuthInitialState, action: PayloadAction<AuthInitialState>) => {
      state.uid = null;
      state.email = null;
      state.username = null;
      state.status = AUTH_STATUS.NOT_AUTHENTICATED
      state.errorMessage = action.payload?.errorMessage
      state.role = null
      state.fullname = null
    },
    checkingCredentials : (state:AuthInitialState) => {
      state.status = AUTH_STATUS.CHECKING
    }
  }
})
export const { login,logout,checkingCredentials } = authSlice.actions