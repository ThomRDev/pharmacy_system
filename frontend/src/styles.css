
*,
*::before,
*::after{
  box-sizing: border-box;
  margin: 0;
  padding: 0;
}

#root{
  min-height: 100vh;
  overflow-x: hidden;
}

.box-shadow{
  box-shadow: 0 5px 5px rgba(0, 0, 0, 0.2);
}
a{
  text-decoration: none;
}
.active .MuiListItemButton-root{
  background-color: #26284687;
}

.loader {
  font-size: 10px;
  width: 1em;
  height: 1em;
  border-radius: 50%;
  position: relative;
  text-indent: -9999em;
  animation: mulShdSpin 1.1s infinite ease;
  transform: translateZ(0);
}
@keyframes mulShdSpin {
  0%,
  100% {
    box-shadow: 0em -2.6em 0em 0em #ffffff, 1.8em -1.8em 0 0em rgba(255,255,255, 0.2), 2.5em 0em 0 0em rgba(255,255,255, 0.2), 1.75em 1.75em 0 0em rgba(255,255,255, 0.2), 0em 2.5em 0 0em rgba(255,255,255, 0.2), -1.8em 1.8em 0 0em rgba(255,255,255, 0.2), -2.6em 0em 0 0em rgba(255,255,255, 0.5), -1.8em -1.8em 0 0em rgba(255,255,255, 0.7);
  }
  12.5% {
    box-shadow: 0em -2.6em 0em 0em rgba(255,255,255, 0.7), 1.8em -1.8em 0 0em #ffffff, 2.5em 0em 0 0em rgba(255,255,255, 0.2), 1.75em 1.75em 0 0em rgba(255,255,255, 0.2), 0em 2.5em 0 0em rgba(255,255,255, 0.2), -1.8em 1.8em 0 0em rgba(255,255,255, 0.2), -2.6em 0em 0 0em rgba(255,255,255, 0.2), -1.8em -1.8em 0 0em rgba(255,255,255, 0.5);
  }
  25% {
    box-shadow: 0em -2.6em 0em 0em rgba(255,255,255, 0.5), 1.8em -1.8em 0 0em rgba(255,255,255, 0.7), 2.5em 0em 0 0em #ffffff, 1.75em 1.75em 0 0em rgba(255,255,255, 0.2), 0em 2.5em 0 0em rgba(255,255,255, 0.2), -1.8em 1.8em 0 0em rgba(255,255,255, 0.2), -2.6em 0em 0 0em rgba(255,255,255, 0.2), -1.8em -1.8em 0 0em rgba(255,255,255, 0.2);
  }
  37.5% {
    box-shadow: 0em -2.6em 0em 0em rgba(255,255,255, 0.2), 1.8em -1.8em 0 0em rgba(255,255,255, 0.5), 2.5em 0em 0 0em rgba(255,255,255, 0.7), 1.75em 1.75em 0 0em #ffffff, 0em 2.5em 0 0em rgba(255,255,255, 0.2), -1.8em 1.8em 0 0em rgba(255,255,255, 0.2), -2.6em 0em 0 0em rgba(255,255,255, 0.2), -1.8em -1.8em 0 0em rgba(255,255,255, 0.2);
  }
  50% {
    box-shadow: 0em -2.6em 0em 0em rgba(255,255,255, 0.2), 1.8em -1.8em 0 0em rgba(255,255,255, 0.2), 2.5em 0em 0 0em rgba(255,255,255, 0.5), 1.75em 1.75em 0 0em rgba(255,255,255, 0.7), 0em 2.5em 0 0em #ffffff, -1.8em 1.8em 0 0em rgba(255,255,255, 0.2), -2.6em 0em 0 0em rgba(255,255,255, 0.2), -1.8em -1.8em 0 0em rgba(255,255,255, 0.2);
  }
  62.5% {
    box-shadow: 0em -2.6em 0em 0em rgba(255,255,255, 0.2), 1.8em -1.8em 0 0em rgba(255,255,255, 0.2), 2.5em 0em 0 0em rgba(255,255,255, 0.2), 1.75em 1.75em 0 0em rgba(255,255,255, 0.5), 0em 2.5em 0 0em rgba(255,255,255, 0.7), -1.8em 1.8em 0 0em #ffffff, -2.6em 0em 0 0em rgba(255,255,255, 0.2), -1.8em -1.8em 0 0em rgba(255,255,255, 0.2);
  }
  75% {
    box-shadow: 0em -2.6em 0em 0em rgba(255,255,255, 0.2), 1.8em -1.8em 0 0em rgba(255,255,255, 0.2), 2.5em 0em 0 0em rgba(255,255,255, 0.2), 1.75em 1.75em 0 0em rgba(255,255,255, 0.2), 0em 2.5em 0 0em rgba(255,255,255, 0.5), -1.8em 1.8em 0 0em rgba(255,255,255, 0.7), -2.6em 0em 0 0em #ffffff, -1.8em -1.8em 0 0em rgba(255,255,255, 0.2);
  }
  87.5% {
    box-shadow: 0em -2.6em 0em 0em rgba(255,255,255, 0.2), 1.8em -1.8em 0 0em rgba(255,255,255, 0.2), 2.5em 0em 0 0em rgba(255,255,255, 0.2), 1.75em 1.75em 0 0em rgba(255,255,255, 0.2), 0em 2.5em 0 0em rgba(255,255,255, 0.2), -1.8em 1.8em 0 0em rgba(255,255,255, 0.5), -2.6em 0em 0 0em rgba(255,255,255, 0.7), -1.8em -1.8em 0 0em #ffffff;
  }
}

/* height */
::-webkit-scrollbar {
  width: 10px;
  height: 8px;
}

/* Track */
::-webkit-scrollbar-track {
  background: #555;
  border-radius:10px;
}

/* Handle */
::-webkit-scrollbar-thumb {
  border-radius: 5px;
  background: #888;
}

/* Handle on hover */
::-webkit-scrollbar-thumb:hover {
  background: #f1f1f1;
}

.pointer{
  cursor: pointer;
}

.mapboxgl-marker{
  cursor: pointer;
}