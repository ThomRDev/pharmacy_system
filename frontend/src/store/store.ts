import { configureStore } from "@reduxjs/toolkit";
import { useDispatch } from "react-redux";
import { inventoryApi } from "./apis/invetoryApi";
import { laboratoryApi } from "./apis/laboratoryApi";
import { authSlice } from "./auth";
import { productSlice } from "./inventory/inventorySlice";
import { laboratorySlice } from "./laboratory/laboratorySlice";
import { employeeSlice } from "./employee/employeeSlice";

export const store = configureStore({
  reducer :{
    auth : authSlice.reducer,
    product : productSlice.reducer,
    employee : employeeSlice.reducer,
    laboratoryStore : laboratorySlice.reducer,
    [inventoryApi.reducerPath] : inventoryApi.reducer,
    [laboratoryApi.reducerPath] : laboratoryApi.reducer,
  },
  middleware: (getDefaultMiddleware) => getDefaultMiddleware().concat(inventoryApi.middleware).concat(laboratoryApi.middleware)
})

export type RootState = ReturnType<typeof store.getState>
export type AppDispatch = typeof store.dispatch
export const useAppDispatch: () => AppDispatch = useDispatch 