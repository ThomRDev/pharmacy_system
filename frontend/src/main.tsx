import React from 'react'
import ReactDOM from 'react-dom/client'
import { Provider } from 'react-redux'
import { BrowserRouter } from 'react-router-dom'
import App from './App'
import { store } from './store'
import "./styles.css"

import mapboxgl from 'mapbox-gl';
mapboxgl.accessToken = 'pk.eyJ1IjoidGhvbXJvbWFuIiwiYSI6ImNsN2kxbGF4cTAzMHo0MW5zZXlhcHZtc2QifQ.G7deqrqIn6bhADb2GXJNKQ';
if(!navigator.geolocation){
  alert( 'Tu navegador no tiene opción de Geolocation' );
  throw new Error('Tu navegador no tiene opción de Geolocation');
}

ReactDOM.createRoot(document.getElementById('root') as HTMLElement).render(
  // <React.StrictMode>
    <Provider store={store}>
      <BrowserRouter>
        <App />
      </BrowserRouter>
    </Provider>
  // </React.StrictMode>
)
