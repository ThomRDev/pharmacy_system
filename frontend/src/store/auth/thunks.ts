import { AnyAction, ThunkAction } from '@reduxjs/toolkit';
import { pharmaryLogin,pharmaryCheckAuth } from '../../helpers/auth';
import { RootState } from '../store';
import { AuthInitialState, checkingCredentials, login, logout } from './authSlice';

interface LoginProps {
  username:string,
  password:string
}

export const startLogin = ({ username,password }:LoginProps):ThunkAction<void,RootState,unknown,AnyAction> => {
  return async (dispatch,getState) => {
    dispatch(checkingCredentials())
    const data = await pharmaryLogin({ username,password })
    if(!data?.ok) return dispatch(logout({  errorMessage : `${data?.statusMsg} : ${data?.error}` } as AuthInitialState ))
    console.log(data)
    const fullname = `${data?.firstname} ${data?.lastname[0]}.`
    dispatch(login( { ...data,fullname, } as AuthInitialState ))
  }
}

export const startCheckingAuthentication = ():ThunkAction<void,RootState,unknown,AnyAction> => {
  return async (dispatch,getState) => {
    if( !localStorage.getItem("token") ) return dispatch(logout({ } as AuthInitialState ))
    const data = await pharmaryCheckAuth(localStorage.getItem("token") as string)
    if(!data?.ok) return dispatch(logout({  errorMessage : `${data?.statusMsg} : ${data?.error}` } as AuthInitialState ))
    const fullname = `${data?.firstname} ${data?.lastname[0]}.`
    dispatch(login( { ...data,fullname } as AuthInitialState ))
  }
}

export const startLogout = ():ThunkAction<void,RootState,unknown,AnyAction> => {
  return async (dispatch,getState) => {
    dispatch(checkingCredentials())
    localStorage.removeItem("token");
    dispatch(logout({ } as AuthInitialState ))
  }
}