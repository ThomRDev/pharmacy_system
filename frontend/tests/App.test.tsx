import { MemoryRouter } from "react-router-dom";
import { configureStore } from "@reduxjs/toolkit";
import {
  act,
  fireEvent,
  getByLabelText,
  getByTestId,
  render,
  screen,
  waitFor,
} from "@testing-library/react";
import { Provider } from "react-redux";
import { useAppDispatch } from "../src/store";
import { authSlice } from "../src/store/auth";
import { notAuthenticatedState } from "./fixtures/authFixtures";
import App from "../src/App";
import { expect } from "@jest/globals";
import "jest";
const mockStartLoginWithUsernamePassword = jest.fn();

interface LoginProps {
  username: string;
  password: string;
}

jest.mock("../src/store/auth/thunks", () => ({
  ...jest.requireActual("../src/store/auth/thunks.ts"),
  startLogin: ({ username, password }: LoginProps) => {
    return () => mockStartLoginWithUsernamePassword({ username, password });
  },
}));

const store = configureStore({
  reducer: {
    auth: authSlice.reducer,
  },
  preloadedState: {
    auth: notAuthenticatedState,
  },
});

jest.mock("../src/store/store.ts", () => ({
  ...jest.requireActual("../src/store/store.ts"),
  useDispatch: () => (fn: any) => fn(),
}));

describe("Test al component <App />", () => {
  beforeEach(() => jest.clearAllMocks());
  test('Deberia hacer match con el snapshot', () => {
    const { container } = render(
      <Provider store={store}>
        <MemoryRouter>
          <App />
        </MemoryRouter>
      </Provider>
    );
    expect(container).toMatchSnapshot();
  })
  test("Debe de renderizar el componente <LoginPage /> cuando no esta autenticado", () => {
    render(
      <Provider store={store}>
        <MemoryRouter>
          <App />
        </MemoryRouter>
      </Provider>
    );
    expect(screen.getAllByText("Login").length).toBeGreaterThanOrEqual(1);
  });
  test("Debe de llamar a startLogin para inicar sesion", async () => {
    const username = "thomtwd";
    const password = "13456";

    render(
      <Provider store={store}>
        <MemoryRouter>
          <App />
        </MemoryRouter>
      </Provider>
    );

    const usernameField = screen.getByLabelText("username");
    fireEvent.change(usernameField, {
      target: { name: "username", value: username },
      preventDefault() {},
    });
    const passwordField = screen.getByLabelText("password");
    fireEvent.change(passwordField, {
      target: { name: "password", value: password },
    });
    const submitFormLogin = screen.getByLabelText("submit-form-login");
    fireEvent.submit(submitFormLogin, { preventDefault() {} });
    await waitFor(() =>
      expect(mockStartLoginWithUsernamePassword).toHaveBeenCalledWith({
        password: "13456",
        username: "thomtwd",
      })
    );
  });
});
