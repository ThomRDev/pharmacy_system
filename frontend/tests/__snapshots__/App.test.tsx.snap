// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test al component <App /> Deberia hacer match con el snapshot 1`] = `
<div>
  <div
    class="MuiGrid-root MuiGrid-container MuiGrid-direction-xs-column css-qaesf5-MuiGrid-root"
  >
    <div
      class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-3 box-shadow css-1ajm08i-MuiGrid-root"
    >
      <form
        aria-label="submit-form-login"
        class="animate__animated animate__fadeIn animate__faster"
      >
        <div
          class="MuiBox-root css-8atqhb"
          style="display: grid; place-items: center;"
        >
          <img
            alt=""
            src=""
            style="height: 150px; display: block; margin: auto;"
          />
        </div>
        <div
          class="MuiGrid-root MuiGrid-container MuiGrid-direction-xs-column css-lx9shp-MuiGrid-root"
        >
          <h6
            class="MuiTypography-root MuiTypography-h6 css-1iyst5o-MuiTypography-root"
          >
            La casa de la salud
          </h6>
          <p
            class="MuiTypography-root MuiTypography-body1 css-k8heu2-MuiTypography-root"
          >
            Login
          </p>
          <div
            class="MuiGrid-root MuiGrid-container css-11lq3yg-MuiGrid-root"
            style="display: flex; margin-top: 1em; gap: 1em;"
          >
            <div
              class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
            >
              <label
                class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-filled MuiFormLabel-colorPrimary css-wkpxuj-MuiFormLabel-root-MuiInputLabel-root"
                data-shrink="false"
                for="username"
                id="username-label"
              >
                Username
              </label>
              <div
                class="MuiInputBase-root MuiFilledInput-root MuiFilledInput-underline MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-1r9izwz-MuiInputBase-root-MuiFilledInput-root"
              >
                <input
                  aria-invalid="false"
                  aria-label="username"
                  autocomplete="off"
                  class="MuiInputBase-input MuiFilledInput-input css-1rv476z-MuiInputBase-input-MuiFilledInput-input"
                  id="username"
                  name="username"
                  placeholder="Admin123"
                  type="text"
                  value=""
                />
              </div>
            </div>
            <div
              class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
            >
              <label
                class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-filled MuiFormLabel-colorPrimary css-wkpxuj-MuiFormLabel-root-MuiInputLabel-root"
                data-shrink="false"
                for="password"
                id="password-label"
              >
                Password
              </label>
              <div
                class="MuiInputBase-root MuiFilledInput-root MuiFilledInput-underline MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-1r9izwz-MuiInputBase-root-MuiFilledInput-root"
              >
                <input
                  aria-invalid="false"
                  aria-label="password"
                  autocomplete="off"
                  class="MuiInputBase-input MuiFilledInput-input css-1rv476z-MuiInputBase-input-MuiFilledInput-input"
                  id="password"
                  name="password"
                  placeholder="***********"
                  type="password"
                  value=""
                />
              </div>
            </div>
          </div>
          <div
            class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 css-1ptpg7m-MuiGrid-root"
            style="margin-top: 0.8em;"
          >
            <div
              class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation0 MuiAlert-root MuiAlert-standardError MuiAlert-standard css-17m4wih-MuiPaper-root-MuiAlert-root"
              role="alert"
            >
              <div
                class="MuiAlert-icon css-1ytlwq5-MuiAlert-icon"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit css-1vooibu-MuiSvgIcon-root"
                  data-testid="ErrorOutlineIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M11 15h2v2h-2zm0-8h2v6h-2zm.99-5C6.47 2 2 6.48 2 12s4.47 10 9.99 10C17.52 22 22 17.52 22 12S17.52 2 11.99 2zM12 20c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8z"
                  />
                </svg>
              </div>
              <div
                class="MuiAlert-message css-1pxa9xg-MuiAlert-message"
              />
            </div>
          </div>
          <div
            class="MuiGrid-root MuiGrid-item css-y0dje9-MuiGrid-root"
          >
            <button
              aria-label="btn-login"
              class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-fullWidth  css-fqffqw-MuiButtonBase-root-MuiButton-root"
              id="btn-login"
              style="background-color: rgb(37, 94, 175); font-weight: bold; padding: 10px 0px;"
              tabindex="0"
              type="submit"
            >
              LoginÂ 
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                data-testid="LoginIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M11 7 9.6 8.4l2.6 2.6H2v2h10.2l-2.6 2.6L11 17l5-5-5-5zm9 12h-8v2h8c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2h-8v2h8v14z"
                />
              </svg>
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
          </div>
        </div>
      </form>
    </div>
  </div>
</div>
`;
